pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,227 @@
+func (p *Plan) fetchLastPushedMaxt() (lastPushedMaxt int64, found bool, err error) {
+	query, err := utils.CreatePrombQuery(p.Mint, p.Maxt, []*labels.Matcher{
+		labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, p.ProgressMetricName),
+		labels.MustNewMatcher(labels.MatchEqual, utils.LabelJob, p.JobName),
+	})
+	if err != nil {
+		return -1, false, fmt.Errorf("fetch-last-pushed-maxt create promb query: %w", err)
+	}
+	result, _, _, err := p.readClient.Read(context.Background(), query)
+	if err != nil {
+		return -1, false, fmt.Errorf("fetch-last-pushed-maxt query result: %w", err)
+	}
+	ts := result.Timeseries
+	if len(ts) == 0 {
+		return -1, false, nil
+	}
+	for _, series := range ts {
+		for i := len(series.Samples) - 1; i >= 0; i-- {
+			if series.Samples[i].Timestamp > lastPushedMaxt {
+				lastPushedMaxt = series.Samples[i].Timestamp
+			}
+		}
+	}
+	if lastPushedMaxt == 0 {
+		return -1, false, nil
+	}
+	return lastPushedMaxt, true, nil
+}
+
+func (p *Plan) DecrementBlockCount() {
+	p.blockCounts.Sub(1)
+}
+
+// ShouldProceed reports whether the fetching process should proceeds further. If any time-range is left to be
+// fetched from the provided time-range, it returns true, else false.
+func (p *Plan) ShouldProceed() bool {
+	return p.lastMaxT < p.Maxt
+}
+
+// Update updates the details of the planner that are dependent on previous fetch stats.
+func (p *Plan) Update(numBytes int) {
+	p.lastNumBytes = numBytes
+}
+
+// NewBlock returns a new block after allocating the time-range for fetch.
+func (p *Plan) NextBlock() (reference *Block, err error, delta int64) {
+	timeDelta := determineTimeDelta(p.lastNumBytes, p.lastTimeRangeDelta)
+	var mint int64
+	if p.lastMaxT == math.MinInt64 {
+		mint = p.Mint
+	} else {
+		mint = p.lastMaxT + 1
+	}
+	maxt := mint + timeDelta
+	if maxt > p.Maxt {
+		maxt = p.Maxt
+	}
+	p.lastMaxT = maxt
+	p.lastTimeRangeDelta = timeDelta
+	bRef, err := p.createBlock(mint, maxt)
+	if err != nil {
+		return nil, fmt.Errorf("next-block: %w", err), timeDelta
+	}
+	return bRef, nil, timeDelta
+}
+
+// createBlock creates a new block and returns reference to the block for faster write and read operations.
+func (p *Plan) createBlock(mint, maxt int64) (reference *Block, err error) {
+	if err = p.validateT(mint, maxt); err != nil {
+		return nil, fmt.Errorf("create-block: %w", err)
+	}
+	id := p.blockCounts.Add(1)
+	timeRangeInMinutes := (maxt - mint) / time.Minute.Milliseconds()
+	percent := float64(maxt-p.Mint) * 100 / float64(p.Maxt-p.Mint)
+	if percent > 100 {
+		percent = 100
+	}
+	baseDescription := fmt.Sprintf("block-%d time-range: %d mins | mint: %d | maxt: %d", id, timeRangeInMinutes, mint, maxt)
+	reference = &Block{
+		id:              id,
+		percent:         percent,
+		pbarInitDetails: baseDescription,
+		pbar: progressbar.NewOptions(
+			6,
+			progressbar.OptionOnCompletion(func() {
+				_, _ = fmt.Fprint(os.Stderr, "\n")
+			}),
+		),
+		mint:    mint,
+		maxt:    maxt,
+		pbarMux: p.pbarMux,
```

Comment:
We need to have a global mutex so that between progress-bars, we can communicate. We also need to make this golbal mux reachable to each block since the actual locking happens in the SetDescription(). Hence, the design.